[WiP] SG-268: Add DBUS API to access control

przemyslawfierek-silvair opened the pull request.

przemyslawfierek-silvair committed:
mesh: Add D-Bus API to 'Access Control List' management to the 'mesh-io-tcpserver'

This patch provides D-Bus API to access control management. From now it
is possible to grant or revoke access for 'uGW' using single D-Bus call.

przemyslawfierek-silvair committed:
mesh: Add D-Bus API to getting connection statistics (for 'mesh-io-tcpserver')

This patch provides D-Bus API to getting information about connection
state. From now it is possible to check if 'uGW' is connected and how
many packets were sent and received.

michallowasrzechonek-silvair commented on mesh/mesh-adapter-tcpserver.c:
```diff
@@ -0,0 +1,516 @@
+	return true;
+}
+
+static bool add_token(json_object *jobj, uint64_t token)
```
I think you can keep the token simply as a uint64 number, without hex encoding.

przemyslawfierek-silvair replied:
`uint64_t` is not supported by JSON-c library. It is possible to store token as `int64_t` or hexstr. Should I change it to store it as `int64_t`?

michallowasrzechonek-silvair replied:
Alright, let's leave it as hexstring then.

michallowasrzechonek-silvair commented on mesh/mesh-io-tcpserver.c:
```diff
@@ -42,17 +42,14 @@
 #include "mesh/mesh-io-api.h"
 #include "mesh/mesh-io-tcpserver.h"
 #include "mesh/silvair-io.h"
+#include "mesh/mesh-adapter-manager.h"
```
I'd rename this to just mesh-adapter.

michallowasrzechonek-silvair commented on mesh/mesh-adapter-tcpserver.c:
```diff
@@ -0,0 +1,533 @@
+	if (!str)
+		return 0;
+
+	data = l_util_from_hexstring(str, &len);
```
There's a poorly named function in util.h: `str2hex`. See how keys are handled in mesh-config-json, might be of interest.

michallowasrzechonek-silvair commented on mesh/mesh-adapter-api.h:
```diff
@@ -0,0 +1,47 @@
+struct mesh_adpt_private;
+
+
+typedef bool (*mesh_adpt_api_reg_dbus_ifaces)(struct l_dbus *bus);
```
These names suck ;)

michallowasrzechonek-silvair replied:
Also, we should be taking adapter as the first argument, following 'object oriented C' conventions.

przemyslawfierek-silvair committed:
mesh: Add D-Bus initialization API to 'mesh-io'

This patch provides API to perform D-Bus initialization to 'mesh-io'.
This should allow to create some D-Bus objects related to 'mesh-io'.

michallowasrzechonek-silvair commented on mesh/mesh-io.c:
```diff
@@ -194,3 +195,13 @@ bool mesh_io_send_cancel(struct mesh_io *io, const uint8_t *pattern,
 	return false;
 }
+
+bool mesh_io_dbus_init(struct l_dbus *dbus)
```
This file implements `mesh_io` object, so all public API functions should take a pointer to "this" object.

You can make the API cleaner by changing the signature to `bool mesh_io_dbus_init(struct mesh_io *io, struct l_dbus *dbus)` and calling this function from `mesh_dbus_init` (passing `mesh.io` as the first argument) instead of calling it from `dbus_init`.

michallowasrzechonek-silvair commented on mesh/silvair-io.c:
```diff
@@ -752,3 +754,10 @@ void silvair_io_destroy(struct silvair_io *io)
 		SSL_free(io->tls_conn);
 }
+
+void silvair_io_close(struct silvair_io *io)
```
Is this related to connection API? I think it *at least* belongs in a separate commit, if not another PR.

przemyslawfierek-silvair replied:
How about create separate PR with this functionality to master and then rebase?

michallowasrzechonek-silvair commented on mesh/tcpserver-acl.h:
```diff
@@ -0,0 +1,46 @@
+
+
+struct tcpserver_acl *tcpserver_acl_new(const char *config_path,
+			on_acl_entry_changed on_acl_entry_created_callback,
```
Meh, just add another argument to the callback to indicate operation.

Also, I don't get the difference between 'created' and 'loaded' - it shouldn't matter if the entry comes from a file or from a dbus call, no? So maybe just a `bool` argument to indicate creation/removal would be enough?

przemyslawfierek-silvair replied:
Let's discuss this.

michallowasrzechonek-silvair commented on mesh/tcpserver-acl.h:
```diff
@@ -0,0 +1,46 @@
+
+void tcpserver_acl_destroy(struct tcpserver_acl *acl);
+
+bool tcpserver_acl_register_dbus_interface(
```
I don't think this needs to be a separate public function. Just rename `tcpserver_acl_create_dbus_object` to `tcpserver_acl_dbus_init` and move interface registration there.

Note that you call these 2 one after another anyway, in `tcpserver_io_dbus_init`, and we don't support multiple instances. This can be changed when/if we get to that point.

michallowasrzechonek-silvair commented on mesh/tcpserver-acl.h:
```diff
@@ -0,0 +1,46 @@
+bool tcpserver_acl_create_dbus_object(struct tcpserver_acl *acl,
+					struct l_dbus *bus, const char *path);
+
+size_t tcpserver_acl_copy_psk(struct tcpserver_acl *acl,
```
I would call it 'tcpserver_acl_psk_get`.

michallowasrzechonek-silvair commented on mesh/tcpserver-acl.c:
```diff
@@ -0,0 +1,552 @@
+		goto fail;
+
+	return entry;
+	fail:
```
indent

michallowasrzechonek-silvair commented on mesh/mesh-io-tcpserver.c:
```diff
@@ -61,24 +63,29 @@ struct mesh_io_private {
 };
 
+struct tcpserver_conn {
+	uint8_t			identity_buf[IDENTITY_BUF_LEN];
```
I'd call it `identity_str`

przemyslawfierek-silvair replied:
`identity` here is kept in the binary representation.

michallowasrzechonek-silvair commented on mesh/mesh-io-tcpserver.c:
```diff
@@ -203,19 +233,28 @@ static void io_error_callback(struct silvair_io *silvair_io)
+	conn = l_queue_find(mesh_io->pvt->conns,
+				match_tcpserver_conn_by_silvair_io, silvair_io);
+	if (conn)
+		conn->io = NULL;
```
Shouldn't we `conn->conn_stat->connected = false` here?

Also, this is the point where you should emit `PropertiesChanged` signal.

przemyslawfierek-silvair replied:
`conn_stat` will provide functions that will emit signals (as we discussed). I've put it here just for test purpose.

michallowasrzechonek-silvair commented on mesh/mesh-io-tcpserver.c:
```diff
@@ -314,51 +351,42 @@ static void tcpserver_io_init_done(void *user_data)
-	// psk is constant length
-	psk_len = 16;
+	if (conn->conn_stat)
+		conn->conn_stat->connected = true;
```
Same: emit `PropertiesChanged` here.

przemyslawfierek-silvair replied:
conn_stat will provide functions that will emit signals (as we discussed). I've put it here just for test purpose.

michallowasrzechonek-silvair commented on mesh/mesh-io-tcpserver.c:
```diff
@@ -314,51 +351,42 @@ static void tcpserver_io_init_done(void *user_data)
-	// psk = k1(dev_key, id, "idpsk")
-	if (!mesh_crypto_k1(pvt->dev_key, id, k1_info, sizeof(k1_info), psk))
-		goto done;
+	conn->io = silvair_io;
```
This part is nice. Exactly what I was hoping for :heart:.

michallowasrzechonek-silvair commented on mesh/mesh-io-tcpserver.c:
```diff
@@ -388,35 +416,53 @@ static bool tls_ctx_init(struct mesh_io_private *pvt)
-	if (!pvt->uuid || len != UUID_LEN) {
-		l_error("Invalid UUID length");
+	if (!hex2str(identity_buf, IDENTITY_BUF_LEN, identity,
+							     sizeof(identity)))
```
I'm somewhat confused if `tcpserver_conn` holds binary or textual representation of peer's identity.

I think we should keep the binary representation in memory and convert strings incoming from TLS handshake only to find a corresponding PSK.

przemyslawfierek-silvair replied:
`tcpserver_conn` keeps binary representation of peer's identity. I've found that in rest of code, binary representation is called with `_buf` postfix, so I've decided to call it similar.

michallowasrzechonek-silvair commented on mesh/mesh-io-tcpserver.c:
```diff
@@ -388,35 +416,53 @@ static bool tls_ctx_init(struct mesh_io_private *pvt)
+	tcpserver_conn_destroy(conn);
+
+	// TODO(prfi):
+	l_queue_remove(mesh_io->pvt->conns, conn);
```
Why is this a `TODO`? Looks fine?

przemyslawfierek-silvair replied:
There is `return false;` to avoid overwriting configuration file (just for test purposes).

michallowasrzechonek-silvair reviewed (changes requested):
Way better!

przemyslawfierek-silvair committed:
mesh: Add 'close' function to 'silvair-io'

This patch adds to 'silvair-io' function to close TCP connection.

michallowasrzechonek-silvair commented on mesh/conn-stat.c:
```diff
@@ -0,0 +1,246 @@
+		return;
+
+	if (connected) {
+		conn_stat->connected = false;
```
Seems redundant?

michallowasrzechonek-silvair commented on mesh/conn-stat.c:
```diff
@@ -0,0 +1,246 @@
+
+	conn_stat->connected = connected;
+
+	l_dbus_property_changed(conn_stat->dbus,
```
formatting ;P

michallowasrzechonek-silvair commented on mesh/mesh-io-tcpserver.c:
```diff
@@ -187,6 +195,12 @@ static void process_rx(struct silvair_io *silvair_io, int8_t rssi,
 		silvair_io_keep_alive_wdt_refresh(silvair_io);
 
+	conn = l_queue_find(mesh_io->pvt->conns,
+			    match_tcpserver_conn_by_silvair_io, silvair_io);
```
formatting

michallowasrzechonek-silvair commented:
```
WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#31: FILE: mesh/mesh-io-api.h:48:
+^Imesh_io_dbus_init_t ^Idbus_init;$

total: 0 errors, 1 warnings, 52 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Commit 9709d88155d7 ("mesh: Add D-Bus initialization API to 'mesh-io'") has style problems, please review.
```

michallowasrzechonek-silvair commented:
Double check the style (see above, do checkpatch on *every* commit!), think about shortening a few function names, and we're ready to merge. Nice work.

michallowasrzechonek-silvair commented:
I think you broke `--io=hci0`:

```
$ sudo ./mesh/bluetooth-meshd --io=hci0
2020-03-10 11:36:10.677       main.c:125  ready_callback             D-Bus ready
2020-03-10 11:36:10.677       main.c:106  request_name_callback      Request name success
2020-03-10 11:36:10.677       mesh.c:263  mesh_init                  Loading node configuration from /var/lib/bluetooth/mesh
...
2020-03-10 11:36:10.792       mesh.c:869  mesh_dbus_init             Added Network Interface on /org/bluez/mesh
2020-03-10 11:36:10.793       main.c:98   mesh_ready_callback        Failed to initialize mesh D-Bus resources
```
